### ISSUE TITLE
Microsoft.Data.SqlClient.SNI not included in Azure Webjob app_data 

### ISSUE THREAD

**Issue by christian-souders** [dotnet/SqlClient#460]
### Describe the bug
SNI.dll files are not present in the app_data directories for Azure Webjob projects that are built in Azure Teams Services using msbuild.  The files are present in the bin directories, but not in the app_data directories.  This leads to runtime errors because the SNI.dll is required.  We have tried upgrading to Microsoft.Data.Sqlclient as suggested in this issue (https://github.com/dotnet/SqlClient/issues/407), but that has not resolved the problem.  We have also tried implementing the workaround to the csproj file as suggested in this issue (https://github.com/dotnet/SqlClient/issues/441).  

Here are some pictures taken from the output of the build that ran in team services that illustrate the problem.  

The bin folder includes the x64 & x86 folders (which include the SNI.dll):
![image](https://user-images.githubusercontent.com/61890247/76242042-39081680-61f3-11ea-8751-d7db408fdb2f.png)

The app_data folder does not include the x64 & x86 folders :
![image](https://user-images.githubusercontent.com/61890247/76242146-694fb500-61f3-11ea-99af-9c77c8dc95bb.png)

### Expected behavior
The SNI.dll file should be present in the app_data directory.

### Further technical details
EntityFrameworkCore version : 3.1.1
Microsoft.Data.SqlClient version: 1.1.1
.NET target: Framework 4.7.2
Nuget version :  5.4.0
Microsoft.Azure.WebJobs version :  v2.3.0

**Comment by cheenamalhotra** [comment]
Hi @christian-souders 

Please upload your repro app with steps that you've tried so we can investigate Azure Web Jobs architecture and provide appropriate targets.

**Comment by jeredm** [comment]
@cheenamalhotra Here is a repository that reproduces the problem. The setup needed is in the README.md file. https://github.com/jeredm/SniWebJobDemo

**Comment by jeredm** [comment]
We are currently using this as a work around:
  ```
<ItemGroup>
    <None Include="$(USERPROFILE)\.nuget\packages\microsoft.data.sqlclient.sni\1.1.1\buildTransitive\net46\x64\SNI.dll">
      <Link>x64\SNI.dll</Link>
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
  </ItemGroup>
  <ItemGroup>
    <None Include="$(USERPROFILE)\.nuget\packages\microsoft.data.sqlclient.sni\1.1.1\buildTransitive\net46\x86\SNI.dll">
      <Link>x86\SNI.dll</Link>
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
  </ItemGroup>
```

The concern with this solution is that the link has a hard coded version and wouldn't adjust with a NuGet version update.

**Comment by cheenamalhotra** [comment]
I tried the repro and I have emulator setup and running, but I'm not an expert on Azure Storage, so can you help me with last 2 steps with code that I can try?

- Create an Azure Queue named sni-webjob-queue
- Add a message to sni-webjob-queue (the content of the message doesn't matter).

**Comment by christian-souders** [comment]
Chenna, 
You can download the azure storage explorer from here : https://azure.microsoft.com/en-us/features/storage-explorer/.  This tool will allow you to create the queue and to upload a message.  

If you are able to get it deployed and running, you will encounter the following runtime error.  I have added bold around the line where the exception is happening :

---- 

Microsoft.Azure.WebJobs.Host.FunctionInvocationException : Exception while executing function: Functions.ProcessQueueMessage ---> System.TypeInitializationException : The type initializer for 'Microsoft.Data.SqlClient.TdsParser' threw an exception. ---> System.TypeInitializationException : The type initializer for 'Microsoft.Data.SqlClient.SNILoadHandle' threw an exception. ---> System.TypeInitializationException : The type initializer for 'Microsoft.Data.SqlClient.SNINativeMethodWrapper' threw an exception. ---> System.ComponentModel.Win32Exception : **Failed to load D:\local\Temp\jobs\continuous\SNIDemo\aaiqpce2.y0d\x64\SNI.dll** ---> System.ComponentModel.Win32Exception : The specified module could not be found
2020-03-11T18:49:52  PID[49540] Information    End of inner exception
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.Data.SqlClient.SNINativeMethodWrapper..cctor()
2020-03-11T18:49:52  PID[49540] Information    End of inner exception
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.Data.SqlClient.SNINativeMethodWrapper.SNIInitialize()
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.Data.SqlClient.SNILoadHandle..ctor()
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.Data.SqlClient.SNILoadHandle..cctor()
2020-03-11T18:49:52  PID[49540] Information    End of inner exception
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.Data.SqlClient.TdsParser..cctor()
2020-03-11T18:49:52  PID[49540] Information    End of inner exception
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.Data.SqlClient.TdsParser..ctor(Boolean MARS,Boolean fAsynchronous)
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.Data.SqlClient.SqlInternalConnectionTds.LoginNoFailover(ServerInfo serverInfo,String newPassword,SecureString newSecurePassword,Boolean redirectedUserInstance,SqlConnectionString connectionOptions,SqlCredential credential,TimeoutTimer timeout)
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.Data.SqlClient.SqlInternalConnectionTds.OpenLoginEnlist(TimeoutTimer timeout,SqlConnectionString connectionOptions,SqlCredential credential,String newPassword,SecureString newSecurePassword,Boolean redirectedUserInstance)
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.Data.SqlClient.SqlInternalConnectionTds..ctor(DbConnectionPoolIdentity identity,SqlConnectionString connectionOptions,SqlCredential credential,Object providerInfo,String newPassword,SecureString newSecurePassword,Boolean redirectedUserInstance,SqlConnectionString userConnectionOptions,SessionData reconnectSessionData,ServerCertificateValidationCallback serverCallback,ClientCertificateRetrievalCallback clientCallback,DbConnectionPool pool,String accessToken,SqlClientOriginalNetworkAddressInfo originalNetworkAddressInfo,Boolean applyTransientFaultHandling,SqlAuthenticationProviderManager sqlAuthProviderManager)
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.Data.SqlClient.SqlConnectionFactory.CreateConnection(DbConnectionOptions options,DbConnectionPoolKey poolKey,Object poolGroupProviderInfo,DbConnectionPool pool,DbConnection owningConnection,DbConnectionOptions userOptions)
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.Data.ProviderBase.DbConnectionFactory.CreatePooledConnection(DbConnectionPool pool,DbConnection owningObject,DbConnectionOptions options,DbConnectionPoolKey poolKey,DbConnectionOptions userOptions)
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.Data.ProviderBase.DbConnectionPool.CreateObject(DbConnection owningObject,DbConnectionOptions userOptions,DbConnectionInternal oldConnection)
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.Data.ProviderBase.DbConnectionPool.UserCreateRequest(DbConnection owningObject,DbConnectionOptions userOptions,DbConnectionInternal oldConnection)
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject,UInt32 waitForMultipleObjectsTimeout,Boolean allowCreate,Boolean onlyOneCheckConnection,DbConnectionOptions userOptions,DbConnectionInternal& connection)
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject,TaskCompletionSource`1 retry,DbConnectionOptions userOptions,DbConnectionInternal& connection)
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.Data.ProviderBase.DbConnectionFactory.TryGetConnection(DbConnection owningConnection,TaskCompletionSource`1 retry,DbConnectionOptions userOptions,DbConnectionInternal oldConnection,DbConnectionInternal& connection)
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.Data.ProviderBase.DbConnectionInternal.TryOpenConnectionInternal(DbConnection outerConnection,DbConnectionFactory connectionFactory,TaskCompletionSource`1 retry,DbConnectionOptions userOptions)
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.Data.SqlClient.SqlConnection.TryOpenInner(TaskCompletionSource`1 retry)
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.Data.SqlClient.SqlConnection.TryOpen(TaskCompletionSource`1 retry)
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.Data.SqlClient.SqlConnection.Open()
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.OpenDbConnection(Boolean errorsExpected)
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.Open(Boolean errorsExpected)
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReader(RelationalCommandParameterObject parameterObject)
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.InitializeReader(DbContext _,Boolean result)
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.Execute[TState,TResult](TState state,Func`3 operation,Func`3 verifySucceeded)
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.MoveNext()
2020-03-11T18:49:52  PID[49540] Information    at System.Linq.Enumerable.SingleOrDefault[TSource](IEnumerable`1 source)
2020-03-11T18:49:52  PID[49540] Information    at SniWebJobDemo.Functions.ProcessQueueMessage(String message,TextWriter log) at D:\a\1\s\SniWebJobDemo\Functions.cs : 23
2020-03-11T18:49:52  PID[49540] Information    at lambda_method(Closure ,Functions ,Object[] )
2020-03-11T18:49:52  PID[49540] Information    at Microsoft.Azure.WebJobs.Host.Executors.VoidMethodInvoker`2.InvokeAsync(TReflected instance,Object[] arguments)
2020-03-11T18:49:52  PID[49540] Information    at async Microsoft.Azure.WebJobs.Host.Executors.FunctionInvoker`2.InvokeAsync[TReflected,TReturnValue](Object instance,Object[] arguments)
2020-03-11T18:49:52  PID[49540] Information    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
2020-03-11T18:49:52  PID[49540] Information    at async Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.InvokeAsync(IFunctionInvoker invoker,ParameterHelper parameterHelper,CancellationTokenSource timeoutTokenSource,CancellationTokenSource functionCancellationTokenSource,Boolean throwOnTimeout,TimeSpan timerInterval,IFunctionInstance instance)
2020-03-11T18:49:52  PID[49540] Information    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
2020-03-11T18:49:52  PID[49540] Information    at async Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.ExecuteWithWatchersAsync(IFunctionInstance instance,ParameterHelper parameterHelper,TraceWriter traceWriter,CancellationTokenSource functionCancellationTokenSource)
2020-03-11T18:49:52  PID[49540] Information    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
2020-03-11T18:49:52  PID[49540] Information    at async Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.ExecuteWithLoggingAsync(??)
2020-03-11T18:49:52  PID[49540] Information    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
2020-03-11T18:49:52  PID[49540] Information    at async Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.ExecuteWithLoggingAsync(??)
2020-03-11T18:49:52  PID[49540] Information    End of inner exception


----

Getting the storage queue up and running will show the runtime error that we are encountering, but if you can at least get the project to build using the YML file included in the repo you will be able to see the app_data directory does not have the SNI dlls as shown in my initial issue.

**Comment by jeredm** [comment]
@cheenamalhotra If you don't want to run the queue locally, you can just create it on Azure. Create the queue at Azure Storage Accounts > <pick or create a storage account> > Queues ("+ Queue" to add the queue) > <select your queue> > "+ Add message" to add a message (set the message test to any string you want). Next, you will need to point the storage account to the your account, so change the value in the `app.config`. The key is found in Azure Storage Accounts > <pick your account> > Access Keys. Let me know if you need any more info.

**Comment by cheenamalhotra** [comment]
Thanks @jeredm 

Appreciate your help! I'm investigating the build part first, since that should capture the targets not getting picked up for SNI.dll, will try deploying once I get the DLLs copied :)

**Comment by cheenamalhotra** [comment]
Hi @jeredm 

I got your problem reproduced and I see when you try to do `/p:DeployOnBuild=true` our SNI.dll copying targets are not yet executed, as they are triggered in the end after "CoreBuild" completes.

![image](https://user-images.githubusercontent.com/13396919/76466284-9099b400-63a4-11ea-908b-a7b03aa8f662.png)

Solution for you, deploy in another MSBuild process after build completes.

**Comment by jeredm** [comment]
Hi @cheenamalhotra,
Thank you for looking into this! I am not sure I understand you solution fully. Are you saying that we should build a second step in the YAML for MSBuild? If you have an example YAML file around, that would be helpful.

**Comment by mccabest** [comment]
@cheenamalhotra What tool are you using to examine the targets in your screen shot above? Looks very handy!

**Comment by cheenamalhotra** [comment]
I use [MSBuild Log Viewer](http://msbuildlog.com/) :)

**Comment by jeredm** [comment]
I added a second MSBuild step with the clean argument set to false and it seems to work now. So in the YAML it looks like this:
```
- task: MSBuild@1
  inputs:
    solution: '**/*.sln'
    platform: '$(BuildPlatform)'
    configuration: '$(BuildConfiguration)'
    msbuildArguments: '/p:DeployOnBuild=true /p:WebPublishMethod=Package /p:SkipInvalidConfigurations=true /p:PackageLocation="$(build.artifactstagingdirectory)\\" /verbosity:normal'
    clean: true

- task: MSBuild@1
  inputs:
    solution: '**/*.sln'
    platform: '$(BuildPlatform)'
    configuration: '$(BuildConfiguration)'
    msbuildArguments: '/p:DeployOnBuild=true /p:WebPublishMethod=Package /p:SkipInvalidConfigurations=true /p:PackageLocation="$(build.artifactstagingdirectory)\\" /verbosity:normal'
    clean: false
```
This seems to work. Thank you again for your help! I keep an eye out for a patch.

**Comment by cheenamalhotra** [comment]
Closing issue as solution is also shipped with v2.0.0
